### ISSUE TITLE
record types

### ISSUE THREAD

**Issue by gavinking** [ceylon/ceylon-spec#745]
Record types are a construct popular in ML-like languages, that are a bit like C-style structs. From our point of view, they are tuples with named elements. An elegant way to add support for record types to Ceylon would be via `shared` parameters of class aliases, for example:

```
class Person(shared String first, shared String last, shared Integer age) => [first, last, age];
```

Then you could write:

```
Person person = Person("Joe", "Bloggs", 32);
String firstName = person.first;
String lastName = person.last;
Integer age = person.age;
```

Where `Person` is just an alias for the type `[String,String,Integer]`, and `person.last` is just sugar for `person[1]`.

But would this feature be useful to us? I dunno. It might be useful for working with APIs that query relational databases. Apart from that, I can't think of much use, since for us classes are just much more powerful.

One use we _do_ have for record types is to encode parameter names to function types. But I don't think the above approach helps us much there. For that problem we would need to be able to write down types like:

```
Callable<Anything, [String first, String last, Integer age]>
```

without needing to define an alias first.

**Comment by lukaseder** [comment]
> It might be useful for working with APIs that query relational databases. Apart from that, I can't think of much use, since for us classes are just much more powerful.

I don't necessarily agree. PL/SQL `RECORD` types, for instance, have been useful for ages als outside of the interaction scope with SQL. Even if Oracle SQL provides a way to implement object-oriented `OBJECT` types, most PL/SQL developers resort to simple `RECORD` types in order to group related variables / parameters throughout an API without creating a "class".

We haven't given up hope that Java 10 will also introduce these sorts of value types, which will be similar to [C#'s struct types](http://msdn.microsoft.com/en-us/library/ah19swz4.aspx). See also [when to use struct](http://stackoverflow.com/q/521298/521799).

**Comment by gavinking** [comment]
@lukaseder I see value types as a different problem. Ceylon was designed from the start to be able to accommodate value types, which is why `Object` is not `Identifiable`. So a value type would just be a class that doesn't satisfy `Identifiable` and obeys certain other rules. I think structs and record types are different things: a struct is a nominal type; a record type is a structural type.

**Comment by lukaseder** [comment]
I agree that records (named or unnamed) don't necessarily have to imply being value types, although if a record is not a value type, then there is not much keeping developers from creating a class instead.

So, in the context of Ceylon, the discussion around record types might evolve into introducing record literals for creating anonymous classes / anonymous structural types.

**Comment by ShalokShalom** [comment]
I am interested in that :D
